internal/terraform/node_resource_plan_instance.go

```diff
@@ -253,7 +255,14 @@ func (n *NodePlannableResourceInstance) managedResourceExecute(ctx EvalContext)
 		}
 
 		if n.importTarget.ID != "" {
-			change.Importing = &plans.Importing{ID: n.importTarget.ID}
+			// change.Importing true means the resource is being imported into
+			// state for the first time, i.e. it was not already in state.
+			//
+			// Import is idempotent, in the sense that importing a resource into
+			// state twice with no further diffs is a no-op non-importing change.
+			if resourceInPriorState == nil {
```

Comment:
This is happening after the import process has already been done, meaning we've already called "ImportResourceState` and `ReadResource` without the context of the configuration. This also leaves open the possibility of other code paths taking action on the fact that this is labeled for import when it in fact is not. If the resource exists in the state, the import process should not begin at all, which I think would mean that we filter from the import targets in the first place.

Reply:
I think that will work in all cases except the following.

Config generation mode. The resource already exists in state, is targeted in an import block, and does not exist in config. If we ignore the import block in this case, the plan will generate a destroy change for the resource. However, the only other option along the lines of this PR would be to generate a non-importing no-op change that nevertheless generates config.

In the unfortunate situation of a resource being already in state and the user nevertheless wanting to generate config, it might be better to error explicitly. That is, when in config generation mode, error if the resource is in state, in the import block, but not in config.

See internal channel for another path the user might take in that situation.